### PR TITLE
  [improvement](mtmv) Support to use current_date() when create async mv

### PR DESCRIPTION
**Pull request by seawinde**
## Proposed changes

Support to use current_date() when create async materialized view by adding
`'enable_nondeterministic_function' = 'true'` in properties when create materialized view.  `enable_nondeterministic_function` is default false.

Here is a example, it will success

>        CREATE MATERIALIZED VIEW mv_name
>        BUILD DEFERRED REFRESH AUTO ON MANUAL
>        DISTRIBUTED BY RANDOM BUCKETS 2
>        PROPERTIES (
>        'replication_num' = '1',
>        'enable_nondeterministic_function' = 'true'
>        )
>        AS
>        SELECT *, unix_timestamp(k3, '%Y-%m-%d %H:%i-%s') from ${tableName} where current_date() > k3;

Note:
`unix_timestamp` is nondeterministic when has no params. it is deterministic when has params which means format column k3 date

another example, it will success

>        CREATE MATERIALIZED VIEW mv_name
>        BUILD DEFERRED REFRESH AUTO ON MANUAL
>        DISTRIBUTED BY RANDOM BUCKETS 2
>        PROPERTIES (
>        'replication_num' = '1',
>        'enable_nondeterministic_function' = 'true'
>        )
>        AS
>        SELECT *, unix_timestamp() from ${tableName} where current_date() > k3;

though `unix_timestamp()` is nondeterministic, we add `'enable_date_nondeterministic_function' = 'true'` in properties

### CODE REVIEW

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Comment by morrySnow** [comment]
support all nondeterministic function ?

**Comment by seawinde** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/NondeterministicFunctionCollector.java [hunk 1]
```diff
@@ -17,31 +17,66 @@
...
+        }
+        return super.visit(plan, collectContext);
+    }
+
+    /**
+     * Function collect context, collected expressions would be added to collectedExpressions,
+     * the expression in whiteFunctionSet would not be collected
+     */
+    public static class FunctionCollectContext {
+        private final List<Expression> collectedExpressions = new LinkedList<>();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/NondeterministicFunctionCollector.java:57]: do not use java's linkedList in any case. it has poor perf with no adv with ArrayList

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/NondeterministicFunctionCollector.java [hunk 2]
```diff
@@ -17,31 +17,66 @@
...
+            if (expressions != null) {
+                this.collectedExpressions.addAll(expressions);
+            }
+        }
+
+        public List<Expression> getCollectedExpressions() {
+            return collectedExpressions;
         }
-        expressions.forEach(expression -> {
-            collectedExpressions.addAll(expression.collect(Nondeterministic.class::isInstance));
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/NondeterministicFunctionCollector.java:43]: i think u could collected isDeterministic == true after move this interface into expression

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/NondeterministicFunctionCollector.java [hunk 3]
```diff
@@ -43,7 +45,14 @@ public Void visit(Plan plan, FunctionCollectContext collectContext) {
         }
         for (Expression expression : expressions) {
             Set<Expression> nondeterministicFunctions =
-                    expression.collect(expr -> !((ExpressionTrait) expr).isDeterministic());
+                    expression.collect(expr -> {
+                        boolean containsNondeterministic = !((ExpressionTrait) expr).isDeterministic();
+                        if (!collectContext.getCollectExpressionTypes().isEmpty()) {
+                            containsNondeterministic &= collectContext.getCollectExpressionTypes().stream()
+                                    .anyMatch(type -> type.isAssignableFrom(expr.getClass()));
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/NondeterministicFunctionCollector.java:52]: why need `type.isAssignableFrom(expr.getClass())`

**Comment by seawinde** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/NondeterministicFunctionCollector.java (hunk 3 above)
Comment (seawinde) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/NondeterministicFunctionCollector.java]: have removed this

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run feut

**Comment by seawinde** [comment]
run buildall

**Review by zddr — APPROVED** [review]
It may be strange to set enable_dondeterministic_function to false through the alter statement after creating a materialized view, but it does not affect the business logic

**Review by morrySnow — APPROVED** [review]

**Pull request merged**